app/controllers/contacts_controller.rb

```diff
@@ -0,0 +1,35 @@
+class ContactsController < ApplicationController
+  before_action do
+    raise "missing provider code" unless params[:provider_code]
+  end
+
+  before_action :build_provider
+  before_action :build_contact
```

Comment:
Did we say we would try avoid before_actions and call the methods inside the actions that need them?

Reply:
We did. Forgot about that. I'll update.

Reply:
Done in 59417fc